runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/streaming/config/StreamingEngineComputationConfigFetcher.java

```diff
@@ -0,0 +1,296 @@
+  }
+
+  private static BackOff defaultConfigBackoff() {
+    return FluentBackoff.DEFAULT
+        .withInitialBackoff(Duration.millis(100))
+        .withMaxBackoff(Duration.standardMinutes(1))
+        .withMaxCumulativeBackoff(Duration.standardMinutes(5))
+        .backoff();
+  }
+
+  private MapTask createMapTask(StreamingComputationConfig computationConfig) {
+    return new MapTask()
+        .setSystemName(computationConfig.getSystemName())
+        .setStageName(computationConfig.getStageName())
+        .setInstructions(computationConfig.getInstructions());
+  }
+
+  @Override
+  public void start() {
+    fetchInitialPipelineGlobalConfig();
+    schedulePeriodicGlobalConfigRequests();
+  }
+
+  @Override
+  public Optional<ComputationConfig> getConfig(String computationId) {
+    Preconditions.checkArgument(
+        !computationId.isEmpty(),
+        "computationId is empty. Cannot fetch computation config without a computationId.");
+    return getComputationConfigInternal(computationId)
+        .flatMap(
+            config -> {
+              Map<String, String> stateNameMap = config.userStepToStateFamilyNameMap();
+              return config
+                  .computationConfig()
+                  .map(
+                      computationConfig ->
+                          ComputationConfig.create(
+                              createMapTask(computationConfig),
+                              computationConfig.getTransformUserNameToStateFamily(),
+                              stateNameMap));
+            });
+  }
+
+  @Override
+  public void stop() {
+    // We have already shutdown or start has not been called.
+    if (globalConfigRefresher.isShutdown() || !hasReceivedGlobalConfig.get()) {
+      return;
+    }
+
+    globalConfigRefresher.shutdown();
+    boolean isTerminated = false;
+    try {
+      isTerminated = globalConfigRefresher.awaitTermination(10, TimeUnit.SECONDS);
+    } catch (InterruptedException e) {
+      LOG.warn("Error occurred shutting down: {}", globalConfigRefresher);
+    }
+    if (!isTerminated) {
+      globalConfigRefresher.shutdownNow();
+    }
+  }
+
+  /**
+   * Initially b Schedules a background thread that periodically sends getConfig requests to
+   * Dataflow Service to obtain the windmill service endpoint.
+   */
+  @SuppressWarnings("FutureReturnValueIgnored")
+  private void schedulePeriodicGlobalConfigRequests() {
+    globalConfigRefresher.scheduleWithFixedDelay(
+        () -> getGlobalConfig().ifPresent(onStreamingConfig),
+        0,
+        globalConfigRefreshPeriodMillis,
+        TimeUnit.MILLISECONDS);
+  }
+
+  /**
+   * Blocks until we have received the initial global config from Dataflow or an exception is
+   * thrown.
+   */
+  private synchronized void fetchInitialPipelineGlobalConfig() {
+    if (!hasReceivedGlobalConfig.get()) {
+      // Get the initial global configuration. This will initialize the windmillServer stub.
+      while (true) {
```

Comment:
nit: seemsl ike you could move !hasReceivedGlobalConfig to the while condition instead

removes some nesting and if we did have multiple callers calling fetchInitialPipelineGlobalConfig it could speed things up.